vault/seal/ocikms/ocikms.go

```diff
@@ -0,0 +1,305 @@
+
+	// Build crypto client
+	kmsCryptoClient, err := keymanagement.NewKmsCryptoClientWithConfigurationProvider(cp, k.cryptoEndpoint)
+	if err != nil {
+		return nil, errwrap.Wrapf("failed creating NewKmsCryptoClientWithConfigurationProvider: {{err}}", err)
+	}
+
+	return &kmsCryptoClient, nil
+}
+
+// Request metadata includes retry policy
+func (k *OCIKMSSeal) getRequestMetadata() common.RequestMetadata {
+	if !KMSShouldRetryOperation {
+		noRetryPolicy := common.NoRetryPolicy()
+		return common.RequestMetadata{
+			RetryPolicy: &noRetryPolicy,
+		}
+	}
+
+	// Only retry for 5xx errors
+	retryOn5xxFunc := func(r common.OCIOperationResponse) bool {
+		return !(r.Error == nil && r.Response.HTTPResponse().StatusCode < 500)
+	}
+	return getRequestMetadataWithCustomizedRetryPolicy(retryOn5xxFunc)
+}
+
+func getRequestMetadataWithCustomizedRetryPolicy(fn func(r common.OCIOperationResponse) bool) common.RequestMetadata {
+	return common.RequestMetadata{
+		RetryPolicy: getExponentialBackoffRetryPolicy(uint(KMSMaximumNumberOfRetries), fn),
+	}
+}
+
+func getExponentialBackoffRetryPolicy(n uint, fn func(r common.OCIOperationResponse) bool) *common.RetryPolicy {
+	// The duration between each retry operation, you might want to wait longer each time the retry fails
+	exponentialBackoff := func(r common.OCIOperationResponse) time.Duration {
+		return time.Duration(math.Pow(float64(2), float64(r.AttemptNumber-1))) * time.Second
+	}
+	policy := common.NewRetryPolicy(n, fn, exponentialBackoff)
+	return &policy
+}
+
+func (k *OCIKMSSeal) SealType() string {
+	return seal.OCIKMS
+}
+
+func (k *OCIKMSSeal) KeyID() string {
+	return k.keyID
+}
+
+func (k *OCIKMSSeal) Init(context.Context) error {
+	return nil
+}
+
+func (k *OCIKMSSeal) Finalize(context.Context) error {
+	return nil
+}
+
+func (k *OCIKMSSeal) Encrypt(ctx context.Context, plaintext []byte) (*physical.EncryptedBlobInfo, error) {
+	defer metrics.MeasureSince(metricEncrypt, time.Now())
+	if plaintext == nil || len(plaintext) == 0 {
+		metrics.SetGauge(metricEncryptFailed, 1)
+		return nil, fmt.Errorf("given plaintext for encryption is nil")
+	}
+
+	// KMS required base64 encrypted plain text before sending to the service
+	b64Plaintext := b64.StdEncoding.EncodeToString(plaintext)
+
+	if k.cryptoClient == nil {
+		metrics.SetGauge(metricEncryptFailed, 1)
+		return nil, fmt.Errorf("nil client")
+	}
+
+	// Build Encrypt Request
+	requestMetadata := k.getRequestMetadata()
+	encryptedDataDetails := keymanagement.EncryptDataDetails{
+		KeyId:     &k.keyID,
+		Plaintext: &b64Plaintext,
+	}
+
+	input := keymanagement.EncryptRequest{
+		EncryptDataDetails: encryptedDataDetails,
+		RequestMetadata:    requestMetadata,
+	}
+	output, err := k.cryptoClient.Encrypt(ctx, input)
+
+	if err != nil {
+		metrics.SetGauge(metricEncryptFailed, 1)
+		return nil, errwrap.Wrapf("error encrypting data: {{err}}", err)
+	}
+	k.logger.Info(fmt.Sprintf("successfully encrypted Vault unseal key. Storing encrypted key to storage"))
```

Comment:
The encrypted data needn't be the Vault unseal key, nor do we know if anything is being stored.